File: Garnet/Distributed/gannetvm.rb
#!/usr/bin/ruby -w

=begin
The aim of this script is to start the gateway and the services as different processes.

The gateway runs on, well, the gateway, and uses the base port number
Every process takes as argument its port number. For "truly" distributed operation every
process will have a different IP address, obtained using the appropriate socket call.
But we must know in advance which process to run on which core.

So we have a flag multi_ip

	if (multi_ip) end else end

If there's only a single IP, the Node ID is added to the base port number
If there are multiple IPs, the Node ID is added to the base IP address
Consequently, we use the Node ID as an argument. And we always increment the base port number with the node id.

For multiple IPs, we will have to use ssh -f (or -n) host command args
For single IPs we simply have system('command args')

To know how many processes to start, and what the identifier would be, we need to parse the SystemConfiguration.

=end
if ENV.has_key?('GANNET_DIR')
    $LOAD_PATH.push("#{ENV['GANNET_DIR']}/Garnet/")
    @@gannet_dir="#{ENV['GANNET_DIR']}"
else
    raise "Please set GANNET_DIR=[path to your Gannet directory]"
end

# essential
DISTR=1
#NSERVICES=1

# config
WORDSZ=64
VERBOSE=1
DEBUG_ALL=1

# useless
VM=1
TO_YAML=0
SEQVM=0
NEW=1
USE_THREADS=0

require 'SBA/ServiceConfiguration'

# I guess what we need to do is open a listening Socket in a separate thread. 
# Then launch the node processes and let them talk to this socket. Once all nodes have sent a ping,
# then we close that socket and launch the gateway

def main()
    parse_opts()
    service_nodes=get_config()

    for k in service_nodes.keys
        v=service_nodes[k];
        node_id=v[0];
        launch_servicenode(node_id);
    end

    launch_gateway();
end

def get_config()

    if @@ymlfile=='SBA.yml'
        @@ymlfile=find_configpath(@@tdc_file)
    end
    puts @@ymlfile
    service_nodes={};

    require 'yaml'

    if File.exist?(@@ymlfile)
        config = YAML.load(File.open(@@ymlfile));
        service_nodes= config['System']['ServiceNodes']
    else
        raise "Configuration file #{@@ymlfile} does not exist."
    end
    return service_nodes
end
def parse_opts()

    require 'optparse'
    opts=OptionParser.new

    help=<<EOH
    Distributed GannnetVM launcher
    Outputs to STDOUT.
    0 [opts] .tdc-file
    -m: multiple IP addresses, for clusters. Default is single-IP.
    -v: verbose (for debugging)
    -d: dry-run (just print the commands);
    -I: Base IP address
    -P: Base port address
    -Y: YAML config file for Gannet (defaults to SBA.yml)    
    -R: Ruby processes   
EOH

    @@base_ip='127.0.0.1'
    opts.on("-I VAL","--ip=VAL",Integer) {|val| @@base_ip=val }
    @@base_port=7188
    opts.on("-P VAL","--port=VAL",Integer) {|val| @@base_port=val }
    @@verbose=0
    opts.on("-v","--verbose") {@@verbose=1;}
    @@rb=''
    opts.on("-R","--ruby") {@@rb='.rb';}
    @@multi_ip=0
    opts.on("-m","--multi") {@@multi_ip=1;}
    @@run=1
    opts.on("-d","--dryrun") {@@run=0;}

    @@ymlfile='SBA.yml'
    opts.on("-Y yml-file","--yml=yml-file",String) {|yml_file| @@ymlfile=yml_file }

    opts.on("-h","--help") {
        puts help
        exit
    }
    args=opts.parse(ARGV)
    if args.length==0
        raise "Please specify a .tdc file."
    end
    @@tdc_file=args.shift
end


def find_configpath(tdc_file)
    tdch=File.open(tdc_file,"r")
    # first read all bytes into a list of Words
    bytewords=[]
    byteword=0
    hwb=0
    tdch.each_byte {|byte|
        byteword+=(byte<<(8*(NBYTES-1-hwb)))
        hwb=hwb+1
        if hwb==NBYTES
            hwb=0
            bytewords.push(byteword)
            byteword=0
        end
    }
    # Then make sense of the list of words:
    # Find number of packets
    npackets = bytewords.shift
    npackets = npackets >> (NBYTES*8-16)
    # For each packet
    for np in 0..npackets-1
        # get the length from the header
        header_word =bytewords.shift
        for phw in 1..2
            bytewords.shift
        end
        length=(header_word & F_Length) >> FS_Length
        # read the payload
        for j in 0..length-1
            bytewords.shift
        end
    end
    # If the bytecode contains the path to the YAML file, it should be in the
# remaining Words
    ymlfile=""
    w=[]
    for word in bytewords
        for i in 0..NBYTES-1
            w[i]=word>>((NBYTES-1-i)*8)&0xFF
            if w[i]!=0
                ymlfile+=w[i].chr()
            end
        end
    end
    return ymlfile
end

def launch_servicenode(node_id)
    res=0;
    command='';
    if @@multi_ip==1
        node_ip=get_node_ip(node_id);
        # First need to scp YAML file to target node I guess?
        command="ssh -f #{node_ip} gannetnode#{@@rb} #{node_id} #{@@ymlfile} #{@@multi_ip}";
    else
        command="gannetnode#{@@rb} #{node_id} #{@@ymlfile} #{@@multi_ip} &";
    end
    puts command if @@verbose;
    if @@run==1
        res=system(command);
    end
    return res;
end

def launch_gateway()
    res=0;
    command='';
    command="gannetgw#{@@rb} #{@@tdc_file} #{@@multi_ip}";
    puts command if @@verbose;
    if @@run==1
        res=system(command);
    end
    return res;
end

def get_node_ip(node_id)
    bytes=@@base_ip.split('.')
    lastbyte=bytes[-1].to_i
    lastbyte+=node_id;
    bytes[-1]="#{lastbyte}"
    return bytes.join('.');
end

main()

File: Garnet/Distributed/gannetvm.pl
#!/usr/bin/perl
use warnings;
use strict;

=pod
The aim of this script is to start the gateway and the services as different processes.

The gateway runs on, well, the gateway, and uses the base port number
Every process takes as argument its port number. For "truly" distributed operation every
process will have a different IP address, obtained using the appropriate socket call.
But we must know in advance which process to run on which core.

So we have a field $base_ip_int

	if ($base_ip_int) {} else {}

If there's only a single IP, the Node ID is added to the base port number
If there are multiple IPs, the Node ID is added to the base IP address
Consequently, we use the Node ID as an argument. And we always increment the base port number with the node id.

For multiple IPs, we will have to use ssh -f (or -n) host command args
For single IPs we simply have system('command args')

To know how many processes to start, and what the identifier would be, we need to parse the SystemConfiguration.

NOTE: on large multicore machines like the ldc, we can run each process on a different core ("set processor affinity") with the taskset command. What I should do is calculate the mask based on the number of nodes and number of cores

=cut


use Data::Dumper;
my $nl="\n";
use Getopt::Std;
use YAML::Syck;
my $gannet_dir='.';
if (exists $ENV{'GANNET_DIR'}) {
    $gannet_dir=$ENV{'GANNET_DIR'};
}

my $blocked=1;
$SIG{"USR1"} = sub {$blocked=0};

my %opts;
getopts( 'dhmRvI:N:P:Y:', \%opts );

if ( $opts{'h'} or !@ARGV ) {
    die "
    Distributed GannnetVM launcher
    Outputs to STDOUT.
    $0 [opts] .tdc-file
    -v: verbose (for debugging)
	-d: dry-run (just print the commands);
	-I: Base IP address, i.e. address the head node for non-shared-mem machines.
	-P: Base port address
    -Y: YAML config file for Gannet (defaults to SBA.yml)    
    -R: Ruby processes
    -N: Number of cores (for shared-mem or NUMA multicore machines)
    \n";
}

my $verbose = $opts{'v'} ? 1 : 0;
my $ymlfile = $opts{'Y'} || 'SBA.yml';
my $run = $opts{'d'}?0:1;
my $rb=$opts{'R'}?'.rb':'';
my $ncores=$opts{'N'}||1;
my $taskset='';

my $base_ip_str='127.0.0.1';
my $base_ip_int= 0;
if ($opts{'I'}) {
 $base_ip_str=$opts{'I'};
 $base_ip_int= ip_address_str_to_int($base_ip_str);
}

my $base_port=$opts{'P'}||7188;

my $tdc_file=shift @ARGV;
$ymlfile=find_configpath($tdc_file);

my %service_nodes=();

if (-e $ymlfile) {
    my $config_href = YAML::Syck::LoadFile($ymlfile);
    my %config = %{$config_href};
    %service_nodes=%{ $config{'System'}{'ServiceNodes'} };
}
my $nnodes=scalar keys %service_nodes;
if ($ncores>1) {
# use taskset. The gateway goes on processor 1
    
}
# Here we launch the barrier and wait for its signal
if ($run) {
    
    my $pid=$$;
    print "barrier.pl $nnodes $pid &\n" if $verbose;
    system("barrier.pl $nnodes $pid &");
#my $pid=`ps | grep barrier`;
    while($blocked) {};
}
print "Got SIGUSR1 from barrier, launch nodes\n" if $verbose;;
$blocked=1;

# Then we launch all the client nodes

for my $k (keys %service_nodes) {
	my $v=$service_nodes{$k};
	my $node_id=$v->[0];
	launch_servicenode($node_id,$base_ip_int);
}

# Here we wait untill all client nodes have talked back to the barrier 
if ($run) {
while($blocked) {};
}
print "Got SIGUSR1 from barrier, launch gateway\n" if $verbose;;

launch_gateway($base_ip_int);
# this only works for single host
system("killall gannetnode$rb");
################################################################################

sub launch_servicenode {
	my $node_id=shift;
	my $base_ip_int=shift;
	my $res=0;
	my $command='';
	if ($base_ip_int) {
		my $node_ip=get_node_ip($node_id);
		# First need to scp YAML file to target node I guess? 
		$command="ssh -f $node_ip gannetnode$rb $node_id $base_ip_int $verbose $ymlfile";
	} else {
    	$taskset=compute_taskset($node_id,$nnodes,$ncores);
		$command=$taskset."gannetnode$rb $node_id $base_ip_int $verbose $ymlfile &";
	}
	print $command,$nl if $verbose;
	if ($run) {
		$res=system($command);
	}
	return $res;
}

sub launch_gateway {
	my $base_ip_int=shift;
	my $res=0;
	my $command='';
	$command="gannetgw$rb base_ip_int $verbose $tdc_file";
	print $command,$nl if $verbose;
	if ($run) {
		$res=system($command);
	}
	return $res;
}

sub get_node_ip {
	my $node_id=shift;
#	my @bytes=split(/\./,$base_ip);
#	$bytes[-1]+=$node_id;
#	return join('.',@bytes);
    return $base_ip_int+$node_id;
}
sub ip_address_str_to_int {
(my $ip_str)=@_;
    my @bytes=split(/\./,$ip_str);
    my $ip_int=($bytes[0] << 24)+($bytes[1] << 16)+($bytes[2] << 8)+$bytes[3]; 
    return $ip_int;
}

sub compute_taskset {
    (my $node_id, my $nnodes, my $ncores)=@_;
    if ($ncores==1) {
        return '';
    } elsif ($ncores>$nnodes) {
        my $mask=sprintf('%x',2**$node_id); 
        return "taskset 0x$mask "; 
    } elsif ($ncores==1) {
        if ($node_id==0) {
            return 'taskset 0x1 ';
        } else {
            return 'taskset 0x2 ';
        }
    } else {
            my $nodes_per_core=int(($nnodes+1)/$ncores); # e.g. if nnodes+1 is 13 and ncores is 4, we have 13/4 = 4, 
            if ($nodes_per_core<$nnodes/$ncores) {
                $nodes_per_core++;
            }
            my $taskset='';
            for my $i (0..$ncores-1) {
                if ($node_id-$nodes_per_core*$i<$nodes_per_core and $node_id-$nodes_per_core*$i>=0) {
                    $taskset= 'taskset 0x'.(2**$i).' '; # so 0,1,2,3 go on core 1;     
                    last;                    
                } else {
                    $taskset= 'taskset 0x'.(2**$ncores).' ';
                }
            }
            return $taskset;    
    }
}

sub find_configpath { 
    (my $tdc_file)=@_;
    my $NBYTES=4;
    my $F_Length=0x01ff0000; 
    my $FS_Length=16;
    if ($tdc_file=~/\.tdc64/) {
        $NBYTES=8;
        $F_Length=0x0000ffff00000000;
        $FS_Length=32;
    }
    open my $TDCH, '<',$tdc_file;
    binmode($TDCH); 
    # Find number of packets
    my $npackets_str='';
    read($TDCH, $npackets_str, $NBYTES, 0);
    my @npackets_bytes=unpack('W'x$NBYTES,$npackets_str);
    my $npackets=$npackets_bytes[0]*8+$npackets_bytes[1];
    print "NPackets: $npackets\n" if $verbose;;
	my $offset=0;
    # For each packet
    for my $np (1..$npackets) {
        # get the length from the header
        my $header_word_str=0; 
        read($TDCH,$header_word_str,$NBYTES,0);
		my @header_word_bytes=unpack('W' x $NBYTES,$header_word_str);
        print "Header: ".join(':',@header_word_bytes)."\n" if $verbose;;
		my $length=0;#($header_word & $F_Length) >> $FS_Length;
		if ($NBYTES==8) {
			$length=$header_word_bytes[2]*256+$header_word_bytes[3];
		} elsif ($NBYTES==4) {
			$length=($header_word_bytes[0]&1)*256+$header_word_bytes[2];
		}
        print "Packet $np length: $length\n" if $verbose;;
        $offset=($length+3)*$NBYTES;
        seek($TDCH,$offset-$NBYTES,1);
    }

    # If the bytecode contains the path to the YAML file, it should be in the
# remaining Words
    my $ymlfile="";
	my $byte='';
    while($ymlfile!~/\.yml/) {
        my $word='';
        read($TDCH,$word,8,0);
#        $offset+=$NBYTES;
        my @t=unpack('a'x8,$word);
		$ymlfile.=join('',@t);
    }
    if (-e $ymlfile) {
    return $ymlfile
    } else {
    $ymlfile=~s/^.*\///;
    $ymlfile="$gannet_dir/SystemConfigurations/$ymlfile";
    if (-e $ymlfile) {
        return $ymlfile;
    } else {
        die "Could not find $ymlfile.\n";
    }
    }
}

